custom eigenvalue normalization for slepc

lindsayad opened the pull request:
Closes #20095

YaqiWang committed:
demonstration of Issue #20095

lindsayad committed:
Testing new formNorm from SLEPc. Getting incorrect eigenvalue

lindsayad committed:
Use our custom bx_norm for pre-scaling

lindsayad committed:
Add tests for eigen nonlinear convergence

YaqiWang commented on test/tests/executioners/eigen_convergence_issue/tests:
```diff
@@ -0,0 +1,29 @@
+    input = 'a.i'
+    expect_out = '4 Nonlinear'
+    absent_out = '5 Nonlinear'
+    cli_args = 'Problem/bx_norm=fluxintegral -eps_power_sign_normalization 0 -eps_tol 1e-10'
```
What is this `-eps_power_sign_normalization`?

lindsayad replied:
Whether to multiply the Bx vector by the sign of the first non-zero element during normalization. This seems to ensure much better convergence for complex eigenvalues, but has poorer convergence for your particular test

YaqiWang replied:
I actually do not quite like this sign of the first non-zero element approach. The dof in a solution vector could be negative although it is very likely the first dof is always positive.

lindsayad replied:
Well now you have the ability to disable it, and we can set that programmatically to be the default in MOOSE if you wish

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/1860547/) on fef33cf : invalidated by @lindsayad

lindsayad committed:
Adjust tolerance for robustness

moosebuild commented:
Job [Python 3.10 (Apptainer)](https://civet.inl.gov/job/1861309/) on 1b0fd0c : invalidated by @lindsayad

lindsayad committed:
Protect against old petsc versions

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1861710/) on 5ebc728 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25204/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1861728/) on 5ebc728 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>191b9d</th>
      <th colspan=4>#25204 5ebc72</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.63%</td>
      <td>85.64%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>92451</td>
      <td>92469</td>
      <td>+18</td>
      <td>23</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15509</td>
      <td>15508</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25204/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25204/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25204/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Python 3.10 (Apptainer)](https://civet.inl.gov/job/1861695/) on 5ebc728 : invalidated by @lindsayad

moosebuild commented:
Job [No optional](https://civet.inl.gov/job/1861724/) on 5ebc728 : invalidated by @lindsayad

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1861718/) on 5ebc728 : invalidated by @lindsayad

lindsayad committed:
Some more test edits

- Use CSVDiff so we can be confident the eigenvalue is good
- Allow 10 or 11 nonlinear iterations for 'test11'
- Add design, issues, and requirements

lindsayad committed:
git mv eigen_convergence_issue eigen_convergence

lindsayad committed:
Loosen tol for csvdiff for legacy eigen

YaqiWang commented on framework/src/problems/EigenProblem.C:
```diff
@@ -637,4 +639,11 @@ EigenProblem::initPetscOutput()
+{
+  mooseAssert(_bx_norm_provided,
+              "We should not get here unless a bx_norm postprocessor has been provided");
+  return getPostprocessorValueByName(getParam<PostprocessorName>("bx_norm"));
```
Can we get the `bx_norm` in the constructor? We can assert whether the name is empty.

lindsayad replied:
I'm not quite sure what you mean. I do get `_bx_norm_provided` in the constructor. I can't get the postprocessor value itself in the constructor since the problem is constructed well before we add postprocessors

YaqiWang replied:
Sorry, I mean get the name in the constructor.

YaqiWang commented:
Will check the branch out and take a closer look.

YaqiWang commented on test/tests/executioners/eigen_convergence/tests:
```diff
@@ -0,0 +1,44 @@
+    input = 'a.i'
+    expect_out = '4 Nonlinear'
+    absent_out = '5 Nonlinear'
+    cli_args = 'Problem/bx_norm=fluxintegral -eps_power_sign_normalization 0 -eps_tol 2e-10'
```
Why is this test not taking 4 free power iterations?

lindsayad replied:
I remember when I tested this on a bigger Griffin problem, the convergence was better without `-eps_power_sign_normalization 0`. I think this may depend on the spectrum of the operator

lindsayad replied:
> Why is this test not taking 4 free power iterations? Oh, sorry, I did not add -eps_tol 2e-10. But I think this should happen automatically. 

This might require another change in SLEPc. I will see

lindsayad replied:
Ok I was able to force the number of free power iterations with just a MOOSE change

YaqiWang commented on test/tests/executioners/eigen_convergence/tests:
```diff
@@ -0,0 +1,44 @@
+    absent_out = '5 Nonlinear'
+    cli_args = 'Problem/bx_norm=fluxintegral -eps_power_sign_normalization 0 -eps_tol 2e-10'
+    petsc_version = '>=3.20.0'
+    requirement = 'The system shall be able to solve an eigenvalue problem using something other than the L2 norm of Bx for normalization and a SLEPc eigenvalue solver.'
```
I noticed that this test takes more linear iterations per newton than `b.i`, any explanation or fix?

lindsayad replied:
we've talked about this in the past. In the SLEPc formulation we just use A to construct the preconditioner but in your old formulation (at least in this test) you include information from B as well. This means that the legacy preconditioning matrix more accurately reflects the Jacobian but you have told me this often leads to the matrix being more poorly conditioned

YaqiWang replied:
Oh. Then this is completely fine because assembling B in real problems can be costly.

It seems that I cannot resolve a conversation, otherwise I can mark some as resolved.

YaqiWang commented on test/tests/executioners/eigen_convergence/a.i:
```diff
@@ -0,0 +1,87 @@
+  solve_type = PJFNK
+  free_power_iterations = 4
+  nl_abs_tol = 2e-10
+  # The first set of power iterations don't seem to respect these options
```
This can be removed if we make the `doFreeNonlinearPowerIterations` to take the exact number of free power iterations.

lindsayad committed:
Get name of bx norm pp in constructor

lindsayad committed:
Always respect required number of free power its

lindsayad committed:
Remove no longer needed tolerance specifications

loganharbour commented on framework/src/problems/EigenProblem.C:
```diff
@@ -637,4 +639,11 @@ EigenProblem::initPetscOutput()
+{
+  mooseAssert(!_bx_norm_name.empty(),
+              "We should not get here unless a bx_norm postprocessor has been provided");
+  return getPostprocessorValueByName(_bx_norm_name);
```
Is there a reason that we do this late instead of early?

loganharbour replied:
I now see your discussion with Yaqi above. This is fine, then.

loganharbour commented on framework/include/problems/EigenProblem.h:
```diff
@@ -275,6 +269,26 @@ class EigenProblem : public FEProblemBase
+
+  /// The name of the Postprocessor providing the Bx norm. This may be empty in which case the
+  /// default L2 norm of Bx will be used as the Bx norm
+  const PostprocessorName _bx_norm_name;
```
If you care (this is not required) - cool opportunity to get used to `std::optional`. Could have a `const std::optional<PostprocessorName> _bx_norm_name`, and then your logic later on is just `operator bool` for `_bx_norm_name`

loganharbour reviewed (approved):
Easy to follow. looks good

GiudGiud commented on framework/src/kernels/MassEigenKernel.C:
```diff
@@ -17,19 +17,23 @@ MassEigenKernel::validParams()
   InputParameters params = EigenKernel::validParams();
   params.addClassDescription("An eigenkernel with weak form $\\lambda(\\psi_i, -u_h)$ where "
```
```suggestion
  params.addClassDescription("An eigenkernel with weak form $\\lambda(\\psi_i, -u_h coeff)$ where "
```

GiudGiud commented on framework/src/kernels/MassEigenKernel.C:
```diff
@@ -17,19 +17,23 @@ MassEigenKernel::validParams()
   InputParameters params = EigenKernel::validParams();
   params.addClassDescription("An eigenkernel with weak form $\\lambda(\\psi_i, -u_h)$ where "
                              "$\\lambda$ is the eigenvalue.");
+  params.addParam<Real>("coefficient", 1.0, "Coefficient of the term");
```
```suggestion
  params.addParam<Real>("coefficient", 1.0, "Coefficient multiplying the term");
```

GiudGiud commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -962,6 +966,15 @@ mooseSlepcEigenFormFunctionAB(SNES /*snes*/, Vec x, Vec Ax, Vec Bx, void * ctx)
+}
+
 void
 attachCallbacksToMat(EigenProblem & eigen_problem, Mat mat, bool eigen)
```
this routine could use comments

GiudGiud reviewed (approved):
looks good to me

lindsayad commented:
ah now I can address small conversations without having to worry about re-requesting review 😄

lindsayad committed:
Use std::optional for _bx_norm_name

lindsayad committed:
Edit param descriptions

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
Add comments to attachCallbacksToMat

lindsayad committed:
Remove what I believe should be unnecessary PetscObjectCompose calls

YaqiWang commented on test/tests/executioners/eigen_convergence/b.i:
```diff
@@ -0,0 +1,73 @@
+  type = NonlinearEigen
+  bx_norm = fluxintegral
+  solve_type = PJFNK
+  free_power_iterations = 4
```
Need to add `output_after_power_iterations = false` so that the tests can pass.

lindsayad replied:
are you talking about the sweep failures on next? Isn't this just about adjusting the tolerance?

YaqiWang replied:
For the failures. This will disable the output after free  power iterations which can be affected by the number of processors.

YaqiWang replied:
It can be fixed by setting l_tol possibly so that each power iteration we get a fixed behavior regardlless of number of processors, but we can just turn this output off.

lindsayad replied:
Oh I understand now. I didn't catch that we were diffing on the "intermediate" result

lindsayad replied:
Then I can possibly remove the `rel_err` setting in the `tests` spec as well 👍
